movement-test/src/main.c

```diff
@@ -64,80 +78,306 @@ UINT8 update_sprite_animation(UINT8 sprite_id, UINT8 *anim, UINT8 direction, UIN
 
 
+    // XXX: not sure this should be here.
+    // Init the sprite used for the player
+    // We have to add an offset because of the way the gb handles sprites
+    move_sprite(PLAYER_SPRITE_ID, player.pos.x + SPRITE_OFFSET_X + scroll_x, player.pos.y + SPRITE_OFFSET_Y + scroll_y);
+
+}
+
+
+void init_player_state() {
+    // XXX: it's wrong to just hardcode the player's position like that.
+    player.pos.x = 80;
+    player.pos.y = 72;
+
+    player.size.w = 8;
+    player.size.h = 8;
+
+    // XXX: still not sure about new_player's precise role.
+    // Is it the properties the player will have during the next frame?
+    // This is not really useful yet.
+    new_player.size.w = player.size.w;
+    new_player.size.h = player.size.h;
+
+    player_direction = PLAYER_DIRECTION_DOWN;
+    player_animation_frame = 0;
+}
+
+
+
+#define MAX(x,y) ((x) < (y) ? (y) : (x))
+#define MIN(x,y) ((x) < (y) ? (x) : (y))
+#define CLAMP(x, minVal, maxVal) MIN(MAX(minVal, x), maxVal)
+#define ABS_CLAMP(x, absMaxVal) CLAMP(x, -absMaxVal, absMaxVal)
+
+
+void handle_dash() {
+  // TODO for dash: use Nolwenn's sprites.
+  // TODO: add a cooldown.
+
+  // If the player is dashing, he can't control his movement anymore
+  // after x iterations of the dashing code, we set is_dashing to 0 and that ends the dash
+  if (is_dashing) {
+    longueur_dash = 12;
+    step1_dash = 4;
+    step2_dash = 7;
+    step3_dash = 9;
+
+    // The length of the dash is reduced if the movement is diagonal.
+    if (dx != 0 && dy != 0){
+      longueur_dash = 8;
+      step1_dash = 2;
+      step2_dash = 4;
+      step3_dash = 6;
+    }
+
+    // End the dash if enough distance has been travelled.
+    if (compteur_dash == longueur_dash) {
+      is_dashing = 0;
+    }
+
+    // The speed increases during the dash
+    if (compteur_dash == step1_dash){
+      dx = dx + dx;
+      dy = dy + dy;
+    }
+    if (compteur_dash == step2_dash){
+      dx = dx + dx;
+      dy = dy + dy;
+    }
+    if (compteur_dash == step3_dash){
+      // The caracter slows down at the end of the dash
+
+      if (dx > 30) dx = dx + 2;
+      if (dx > 0 && dx < 30) dx = dx -2;
+      if (dy > 30) dy = dy + 2;
+      if (dy > 0 && dx < 30) dy = dy -2;
+    }
+
+    //move_sprite(PLAYER_SPRITE_ID, player.pos.x, player.pos.y);
+    compteur_dash += 1;
+  }
+}
+
+
+void handle_collisions() {
+  total_diff.dx = dx;
+  total_diff.dy = dy;
+
+  while (total_diff.dx != 0 || total_diff.dy != 0) {
+    // Split movement into submoves (to avoid going past obstacles)
+
+    // XXX: the generated assembly for this lines is ridiculously
+    // long IMO (~104 instructions per line!).
+    current_diff.dx = ABS_CLAMP(total_diff.dx, MAX_MOVE_X);
+
+    total_diff.dx -= current_diff.dx;
+    new_player.pos.x = player.pos.x + current_diff.dx;
+
+
+    // Player's position using tilemap coordinates
+    effective_x = (new_player.pos.x >> 3);
+    effective_y = (player.pos.y >> 3);
```

Comment:
Why is the effective `x` coordinate based on `new_player` while `y` is based on `player`?...

Reply:
`new_player.pos.x` contains the new x pos whereas `new_player.pos.y` is undefined. Besides, since we're only testing on the X axis, we only need to update the x pos first.

Reply:
Ah, I see. Thanks!